### ISSUE TITLE
NullPointerException in XMLDatabase when new <enforced> element is missing

### ISSUE THREAD

**Issue by lastcmaster** [jOOQ/jOOQ#9916]
### Expected behavior and actual behavior:
TableConstraint::enforced should not be null.
### Steps to reproduce the problem (if possible, create an MCVE: https://github.com/jOOQ/jOOQ-mcve):
Running the following command line on the attached files with <enforced> commented results in an NPE:

```
java -cp ${REPO}/org/jooq/pro-java-8/jooq/3.13.1/jooq-3.13.1.jar:${REPO}/org/jooq/pro-java-8/jooq-meta/3.13.1/jooq-meta-3.13.1.jar:${REPO}/org/jooq/pro-java-8/jooq-codegen/3.13.1/jooq-codegen-3.13.1.jar:${REPO}/org/reactivestreams/reactive-streams/1.0.2/reactive-streams-1.0.2.jar org.jooq.codegen.GenerationTool codegen.xml
```

```
glz@Goran-MBP main % java -Djooq.codegen.logging=TRACE -cp /Users/glz/.m2/repository/org/jooq/pro-java-8/jooq/3.13.1/jooq-3.13.1.jar:/Users/glz/.m2/repository/org/jooq/pro-java-8/jooq-meta/3.13.1/jooq-meta-3.13.1.jar:/Users/glz/.m2/repository/org/jooq/pro-java-8/jooq-codegen/3.13.1/jooq-codegen-3.13.1.jar:/Users/glz/.m2/repository/org/reactivestreams/reactive-streams/1.0.2/reactive-streams-1.0.2.jar org.jooq.codegen.GenerationTool src/main/resources/codegen.xml
Mar 04, 2020 11:14:51 AM org.jooq.tools.JooqLogger info
INFO: Initialising properties  : src/main/resources/codegen.xml
Mar 04, 2020 11:14:51 AM org.jooq.tools.JooqLogger warn
WARNING: org.xml.sax.SAXParseException; lineNumber: 20; columnNumber: 133; cvc-elt.1: Cannot find the declaration of element 'configuration'.
Mar 04, 2020 11:14:51 AM org.jooq.tools.JooqLogger info
INFO: No <inputCatalog/> was provided. Generating ALL available catalogs instead.
Mar 04, 2020 11:14:52 AM org.jooq.tools.JooqLogger info
INFO: License parameters       
Mar 04, 2020 11:14:52 AM org.jooq.tools.JooqLogger info
INFO: ----------------------------------------------------------
Mar 04, 2020 11:14:52 AM org.jooq.tools.JooqLogger info
INFO:   Thank you for using jOOQ and jOOQ's code generator
Mar 04, 2020 11:14:52 AM org.jooq.tools.JooqLogger info
INFO:                          
Mar 04, 2020 11:14:52 AM org.jooq.tools.JooqLogger info
INFO: Database parameters      
Mar 04, 2020 11:14:52 AM org.jooq.tools.JooqLogger info
INFO: ----------------------------------------------------------
Mar 04, 2020 11:14:52 AM org.jooq.tools.JooqLogger info
INFO:   dialect                : DEFAULT
Mar 04, 2020 11:14:52 AM org.jooq.tools.JooqLogger info
INFO:   URL                    : 
Mar 04, 2020 11:14:52 AM org.jooq.tools.JooqLogger info
INFO:   target dir             : /Users/glz/projects/depot/coordinator/eventfilter/main/target/generated-sources/jooq/java
Mar 04, 2020 11:14:52 AM org.jooq.tools.JooqLogger info
INFO:   target package         : com.ismobile.coordsrv.database
Mar 04, 2020 11:14:52 AM org.jooq.tools.JooqLogger info
INFO:   includes               : [.*]
Mar 04, 2020 11:14:52 AM org.jooq.tools.JooqLogger info
INFO:   excludes               : []
Mar 04, 2020 11:14:52 AM org.jooq.tools.JooqLogger info
INFO:   includeExcludeColumns  : false
Mar 04, 2020 11:14:52 AM org.jooq.tools.JooqLogger info
INFO: ----------------------------------------------------------
Mar 04, 2020 11:14:52 AM org.jooq.tools.JooqLogger info
INFO:                          
Mar 04, 2020 11:14:52 AM org.jooq.tools.JooqLogger info
INFO: JavaGenerator parameters 
Mar 04, 2020 11:14:52 AM org.jooq.tools.JooqLogger info
INFO: ----------------------------------------------------------
Mar 04, 2020 11:14:52 AM org.jooq.tools.JooqLogger info
INFO:   annotations (generated): false
Mar 04, 2020 11:14:52 AM org.jooq.tools.JooqLogger info
INFO:   annotations (JPA: any) : false
Mar 04, 2020 11:14:52 AM org.jooq.tools.JooqLogger info
INFO:   annotations (JPA: version): 
Mar 04, 2020 11:14:52 AM org.jooq.tools.JooqLogger info
INFO:   annotations (validation): false
Mar 04, 2020 11:14:52 AM org.jooq.tools.JooqLogger info
INFO:   comments               : true
Mar 04, 2020 11:14:52 AM org.jooq.tools.JooqLogger info
INFO:   comments on attributes : true
Mar 04, 2020 11:14:52 AM org.jooq.tools.JooqLogger info
INFO:   comments on catalogs   : true
Mar 04, 2020 11:14:52 AM org.jooq.tools.JooqLogger info
INFO:   comments on columns    : true
Mar 04, 2020 11:14:52 AM org.jooq.tools.JooqLogger info
INFO:   comments on keys       : true
Mar 04, 2020 11:14:52 AM org.jooq.tools.JooqLogger info
INFO:   comments on links      : true
Mar 04, 2020 11:14:52 AM org.jooq.tools.JooqLogger info
INFO:   comments on packages   : true
Mar 04, 2020 11:14:52 AM org.jooq.tools.JooqLogger info
INFO:   comments on parameters : true
Mar 04, 2020 11:14:52 AM org.jooq.tools.JooqLogger info
INFO:   comments on queues     : true
Mar 04, 2020 11:14:52 AM org.jooq.tools.JooqLogger info
INFO:   comments on routines   : true
Mar 04, 2020 11:14:52 AM org.jooq.tools.JooqLogger info
INFO:   comments on schemas    : true
Mar 04, 2020 11:14:52 AM org.jooq.tools.JooqLogger info
INFO:   comments on sequences  : true
Mar 04, 2020 11:14:52 AM org.jooq.tools.JooqLogger info
INFO:   comments on tables     : true
Mar 04, 2020 11:14:52 AM org.jooq.tools.JooqLogger info
INFO:   comments on udts       : true
Mar 04, 2020 11:14:52 AM org.jooq.tools.JooqLogger info
INFO:   sources                : true
Mar 04, 2020 11:14:52 AM org.jooq.tools.JooqLogger info
INFO:   sources on views       : true
Mar 04, 2020 11:14:52 AM org.jooq.tools.JooqLogger info
INFO:   daos                   : false
Mar 04, 2020 11:14:52 AM org.jooq.tools.JooqLogger info
INFO:   deprecated code        : false
Mar 04, 2020 11:14:52 AM org.jooq.tools.JooqLogger info
INFO:   global references (any): true
Mar 04, 2020 11:14:52 AM org.jooq.tools.JooqLogger info
INFO:   global references (catalogs): true
Mar 04, 2020 11:14:52 AM org.jooq.tools.JooqLogger info
INFO:   global references (keys): true
Mar 04, 2020 11:14:52 AM org.jooq.tools.JooqLogger info
INFO:   global references (links): true
Mar 04, 2020 11:14:52 AM org.jooq.tools.JooqLogger info
INFO:   global references (queues): true
Mar 04, 2020 11:14:52 AM org.jooq.tools.JooqLogger info
INFO:   global references (routines): true
Mar 04, 2020 11:14:52 AM org.jooq.tools.JooqLogger info
INFO:   global references (schemas): true
Mar 04, 2020 11:14:52 AM org.jooq.tools.JooqLogger info
INFO:   global references (sequences): true
Mar 04, 2020 11:14:52 AM org.jooq.tools.JooqLogger info
INFO:   global references (tables): true
Mar 04, 2020 11:14:52 AM org.jooq.tools.JooqLogger info
INFO:   global references (udts): true
Mar 04, 2020 11:14:52 AM org.jooq.tools.JooqLogger info
INFO:   indexes                : true
Mar 04, 2020 11:14:52 AM org.jooq.tools.JooqLogger info
INFO:   instance fields        : true
Mar 04, 2020 11:14:52 AM org.jooq.tools.JooqLogger info
INFO:   interfaces             : false
Mar 04, 2020 11:14:52 AM org.jooq.tools.JooqLogger info
INFO:   interfaces (immutable) : false
Mar 04, 2020 11:14:52 AM org.jooq.tools.JooqLogger info
INFO:   javadoc                : true
Mar 04, 2020 11:14:52 AM org.jooq.tools.JooqLogger info
INFO:   keys                   : true
Mar 04, 2020 11:14:52 AM org.jooq.tools.JooqLogger info
INFO:   links                  : true
Mar 04, 2020 11:14:52 AM org.jooq.tools.JooqLogger info
INFO:   pojos                  : false
Mar 04, 2020 11:14:52 AM org.jooq.tools.JooqLogger info
INFO:   pojos (immutable)      : false
Mar 04, 2020 11:14:52 AM org.jooq.tools.JooqLogger info
INFO:   queues                 : true
Mar 04, 2020 11:14:52 AM org.jooq.tools.JooqLogger info
INFO:   records                : true
Mar 04, 2020 11:14:52 AM org.jooq.tools.JooqLogger info
INFO:   routines               : true
Mar 04, 2020 11:14:52 AM org.jooq.tools.JooqLogger info
INFO:   sequences              : true
Mar 04, 2020 11:14:52 AM org.jooq.tools.JooqLogger info
INFO:   sequenceFlags          : true
Mar 04, 2020 11:14:52 AM org.jooq.tools.JooqLogger info
INFO:   table-valued functions : true
Mar 04, 2020 11:14:52 AM org.jooq.tools.JooqLogger info
INFO:   tables                 : true
Mar 04, 2020 11:14:52 AM org.jooq.tools.JooqLogger info
INFO:   udts                   : true
Mar 04, 2020 11:14:52 AM org.jooq.tools.JooqLogger info
INFO:   relations              : true
Mar 04, 2020 11:14:52 AM org.jooq.tools.JooqLogger info
INFO: ----------------------------------------------------------
Mar 04, 2020 11:14:52 AM org.jooq.tools.JooqLogger info
INFO:                          
Mar 04, 2020 11:14:52 AM org.jooq.tools.JooqLogger info
INFO: Generation remarks       
Mar 04, 2020 11:14:52 AM org.jooq.tools.JooqLogger info
INFO: ----------------------------------------------------------
Mar 04, 2020 11:14:52 AM org.jooq.tools.JooqLogger info
INFO:                          
Mar 04, 2020 11:14:52 AM org.jooq.tools.JooqLogger info
INFO: ----------------------------------------------------------
Mar 04, 2020 11:14:52 AM org.jooq.tools.JooqLogger info
INFO: Generating catalogs      : Total: 1
Mar 04, 2020 11:14:52 AM org.jooq.tools.JooqLogger info
INFO: Reading from: src/main/resources/coorddb.xml [*]
Mar 04, 2020 11:14:52 AM org.jooq.tools.JooqLogger info
INFO: ARRAYs fetched           : 0 (0 included, 0 excluded)
Mar 04, 2020 11:14:52 AM org.jooq.tools.JooqLogger info
INFO: Enums fetched            : 0 (0 included, 0 excluded)
Mar 04, 2020 11:14:52 AM org.jooq.tools.JooqLogger info
INFO: Packages fetched         : 0 (0 included, 0 excluded)
Mar 04, 2020 11:14:52 AM org.jooq.tools.JooqLogger info
INFO: Routines fetched         : 0 (0 included, 0 excluded)
Mar 04, 2020 11:14:52 AM org.jooq.tools.JooqLogger info
INFO: Sequences fetched        : 0 (0 included, 0 excluded)
Mar 04, 2020 11:14:52 AM org.jooq.tools.JooqLogger info
INFO: Tables fetched           : 1 (1 included, 0 excluded)
Mar 04, 2020 11:14:52 AM org.jooq.tools.JooqLogger info
INFO: No schema version is applied for catalog . Regenerating.
Mar 04, 2020 11:14:52 AM org.jooq.tools.JooqLogger info
INFO:                          
Mar 04, 2020 11:14:52 AM org.jooq.tools.JooqLogger info
INFO: Generating catalog       : DefaultCatalog.java
Mar 04, 2020 11:14:52 AM org.jooq.tools.JooqLogger info
INFO: ==========================================================
Mar 04, 2020 11:14:52 AM org.jooq.tools.JooqLogger info
INFO: Generating schemata      : Total: 1
Mar 04, 2020 11:14:52 AM org.jooq.tools.JooqLogger info
INFO: No schema version is applied for schema COORDDB_3_27. Regenerating.
Mar 04, 2020 11:14:52 AM org.jooq.tools.JooqLogger info
INFO: Generating schema        : Coorddb_3_27.java
Mar 04, 2020 11:14:52 AM org.jooq.tools.JooqLogger info
INFO: ----------------------------------------------------------
Mar 04, 2020 11:14:52 AM org.jooq.tools.JooqLogger info
INFO: UDTs fetched             : 0 (0 included, 0 excluded)
Mar 04, 2020 11:14:52 AM org.jooq.tools.JooqLogger info
INFO: Generating tables        
Mar 04, 2020 11:14:52 AM org.jooq.tools.JooqLogger error
SEVERE: Error while generating table COORDDB_3_27.T_EVENT
java.lang.RuntimeException: java.lang.RuntimeException: java.lang.NullPointerException
	at org.jooq.meta.AbstractDatabase.onError(AbstractDatabase.java:2557)
	at org.jooq.meta.AbstractDatabase.getRelations(AbstractDatabase.java:2016)
	at org.jooq.meta.DefaultColumnDefinition.getPrimaryKey(DefaultColumnDefinition.java:91)
	at org.jooq.meta.AbstractTableDefinition.getPrimaryKey(AbstractTableDefinition.java:86)
	at org.jooq.codegen.JavaGenerator.generateTable(JavaGenerator.java:3880)
	at org.jooq.codegen.JavaGenerator.generateTable(JavaGenerator.java:3874)
	at org.jooq.codegen.JavaGenerator.generateTables(JavaGenerator.java:3861)
	at org.jooq.codegen.JavaGenerator.generate(JavaGenerator.java:519)
	at org.jooq.codegen.JavaGenerator.generate(JavaGenerator.java:473)
	at org.jooq.codegen.JavaGenerator.generate(JavaGenerator.java:394)
	at org.jooq.codegen.GenerationTool.run0(GenerationTool.java:845)
	at org.jooq.codegen.GenerationTool.run(GenerationTool.java:222)
	at org.jooq.codegen.GenerationTool.generate(GenerationTool.java:217)
	at org.jooq.codegen.GenerationTool.main(GenerationTool.java:189)
Caused by: java.lang.RuntimeException: java.lang.NullPointerException
	at org.jooq.meta.AbstractDatabase.onError(AbstractDatabase.java:2557)
	at org.jooq.meta.AbstractDatabase.getRelations0(AbstractDatabase.java:2304)
	at org.jooq.meta.AbstractDatabase$13.run(AbstractDatabase.java:2019)
	at org.jooq.meta.AbstractDatabase.onError(AbstractDatabase.java:2547)
	... 13 more
Caused by: java.lang.NullPointerException
	at org.jooq.meta.xml.XMLDatabase.loadPrimaryKeys(XMLDatabase.java:360)
	at org.jooq.meta.AbstractDatabase$17.run(AbstractDatabase.java:2307)
	at org.jooq.meta.AbstractDatabase.onError(AbstractDatabase.java:2547)
	... 16 more

Mar 04, 2020 11:14:52 AM org.jooq.tools.JooqLogger info
INFO: Tables generated         : Total: 249.842ms
Mar 04, 2020 11:14:52 AM org.jooq.tools.JooqLogger info
INFO: Embeddables fetched      : 0 (0 included, 0 excluded)
Mar 04, 2020 11:14:52 AM org.jooq.tools.JooqLogger info
INFO: Generating table references
Mar 04, 2020 11:14:52 AM org.jooq.tools.JooqLogger info
INFO: Table refs generated     : Total: 254.922ms, +5.079ms
Mar 04, 2020 11:14:52 AM org.jooq.tools.JooqLogger info
INFO: Generating Keys          
Mar 04, 2020 11:14:52 AM org.jooq.tools.JooqLogger info
INFO: Skipping empty keys      
Mar 04, 2020 11:14:52 AM org.jooq.tools.JooqLogger info
INFO: Generating Indexes       
Mar 04, 2020 11:14:52 AM org.jooq.tools.JooqLogger info
INFO: Indexes fetched          : 0 (0 included, 0 excluded)
Mar 04, 2020 11:14:52 AM org.jooq.tools.JooqLogger info
INFO: Skipping empty indexes   
Mar 04, 2020 11:14:52 AM org.jooq.tools.JooqLogger info
INFO: Generating table records 
Mar 04, 2020 11:14:52 AM org.jooq.tools.JooqLogger info
INFO: Generating record        : TEventRecord.java
Mar 04, 2020 11:14:52 AM org.jooq.tools.JooqLogger info
INFO: Forcing type             : COORDDB_3_27.T_EVENT.TIMESTAMP to <userType>java.time.Instant</userType><converter>com.ismobile.hpdoc.eventfilter.bindings.TimestampBinding</converter><includeExpression>.*TIMESTAMP</includeExpression><includeTypes>.*</includeTypes><nullability>ALL</nullability><objectType>ALL</objectType>
Mar 04, 2020 11:14:52 AM org.jooq.tools.JooqLogger info
INFO: Forcing type             : COORDDB_3_27.T_EVENT.TIMESTAMP to <userType>java.time.Instant</userType><converter>com.ismobile.hpdoc.eventfilter.bindings.TimestampBinding</converter><includeExpression>.*TIMESTAMP</includeExpression><includeTypes>.*</includeTypes><nullability>ALL</nullability><objectType>ALL</objectType>
Mar 04, 2020 11:14:52 AM org.jooq.tools.JooqLogger info
INFO: Table records generated  : Total: 307.572ms, +52.65ms
Mar 04, 2020 11:14:52 AM org.jooq.tools.JooqLogger info
INFO: Domains fetched          : 0 (0 included, 0 excluded)
Mar 04, 2020 11:14:52 AM org.jooq.tools.JooqLogger info
INFO: Generation finished: COORDDB_3_27: Total: 308.37ms, +0.798ms
Mar 04, 2020 11:14:52 AM org.jooq.tools.JooqLogger info
INFO:                          
Mar 04, 2020 11:14:52 AM org.jooq.tools.JooqLogger info
INFO: Removing excess files    
```

Uncommenting the element removes the NPE.
[eventcodegen.tar.gz](https://github.com/jOOQ/jOOQ/files/4286683/eventcodegen.tar.gz)

Line 360 in XMLDatabse.java:

```java
relations.addUniqueKey(key, table, table.getColumn(columnName), tc.isEnforced());

TableConstraint tc ...
```

TableConstraint implements the flag enforced as an instance of Boolean so it defaults to null as it is not initialised. If it is not included in the XML, it has no default value.

```java
tc.isEnforced()
    /**
     * Gets the value of the enforced property.
     *
     * @return
     *     possible object is
     *     {@link Boolean }
     *
     */
    public Boolean isEnforced() {
        return enforced;
    }
```

 but in DefaultRelations.java:

```java
    public void addPrimaryKey(String keyName, TableDefinition table, ColumnDefinition column, boolean enforced) {
```

so the Boolean is mapped to a boolean and thus NPE.

### Versions:

- jOOQ: 3.13.1
- Java: 8
- Database (include vendor): XML
- OS: macOS 10.15.3
- JDBC Driver (include name if inofficial driver): None

**Comment by lukaseder** [comment]
Thanks a lot for your detailed bug report. I think there's a second chance of a NPE occurring here. There's no guarantee that a `TableConstraint` item is even present for any given constraint name. That would of course be an invalid XML file, but we should still be more lenient with respect to that.

**Comment by lukaseder** [comment]
Hmm, no that can't happen. We're calling `XMLDatabase::keyColumnUsage` first, which doesn't return a constraint without a `<table_constraint>` entry. So only the explanation you've provided is valid.

**Comment by lukaseder** [comment]
Fixed for jOOQ 3.14.0. Backport scheduled for 3.12.2 (#9919).

A workaround is to place `<enforced>true</enforced>` in each `<table_constraint>`
